File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/12-BST/LeetCode/LeetCode-0098.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private boolean validate(TreeNode root, long leftVal, long rightVal){
        if (root == null) return true;
        if (root.val < leftVal || root.val > rightVal) return false;
        if (!validate(root.left, leftVal, root.val-1L)) return false;
        return validate(root.right, root.val+1L, rightVal);
    } // end-validate

    public boolean isValidBST(TreeNode root) {
        return validate(root, Long.MIN_VALUE, Long.MAX_VALUE);        
    }
}

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/12-BST/LeetCode/LeetCode-0098.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    bool validate(TreeNode *root, int64_t leftVal, int64_t rightVal){
        if (root == nullptr) return true;
        if (root->val < leftVal || root->val > rightVal) return false;
        if (!validate(root->left, leftVal, root->val-1L)) return false;
        return validate(root->right, root->val+1L, rightVal);
    } // end-validate

public:
    bool isValidBST(TreeNode* root) {
        return validate(root, INT64_MIN, INT64_MAX);
    }
};